patroni/quorum.py

```diff
@@ -0,0 +1,429 @@
+        """Handle cases when set of transitions produces on previous run was interrupted.
+
+        :yields: transitions as :class:`Transition` objects.
+        """
+        if self.sync < self.voters:
+            logger.debug("Case 1: synchronous_standby_names subset of DCS state")
+            # Case 1: voters is superset of sync nodes. In the middle of changing voters (quorum).
+            # Evict  dead nodes from voters that are not being synced.
+            remove_from_voters = self.voters - (self.sync | self.active)
+            if remove_from_voters:
+                yield from self.quorum_update(
+                    quorum=len(self.voters) - len(remove_from_voters) - self.numsync,
+                    voters=CaseInsensitiveSet(self.voters - remove_from_voters),
+                    adjust_quorum=not (self.sync - self.active))
+            # Start syncing to nodes that are in voters and alive
+            add_to_sync = (self.voters & self.active) - self.sync
+            if add_to_sync:
+                yield from self.sync_update(self.numsync, CaseInsensitiveSet(self.sync | add_to_sync))
+        elif self.sync > self.voters:
+            logger.debug("Case 2: synchronous_standby_names superset of DCS state")
+            # Case 2: sync is superset of voters nodes. In the middle of changing replication factor (sync).
+            # Add to voters nodes that are already synced and active
+            add_to_voters = (self.sync - self.voters) & self.active
+            if add_to_voters:
+                voters = CaseInsensitiveSet(self.voters | add_to_voters)
+                yield from self.quorum_update(len(voters) - self.numsync, voters)
+            # Remove from sync nodes that are dead
+            remove_from_sync = self.sync - self.voters
+            if remove_from_sync:
+                yield from self.sync_update(
+                    numsync=min(self.numsync, len(self.sync) - len(remove_from_sync)),
+                    sync=CaseInsensitiveSet(self.sync - remove_from_sync))
+
+        # After handling these two cases voters and sync must match.
+        assert self.voters == self.sync
+
+        safety_margin = self.quorum + min(self.numsync, self.numsync_confirmed) - len(self.voters | self.sync)
+        if safety_margin > 0:  # In the middle of changing replication factor.
+            if self.numsync > self.sync_wanted:
+                logger.debug('Case 3: replication factor is bigger than needed')
+                yield from self.sync_update(max(self.sync_wanted, len(self.voters) - self.quorum), self.sync)
+            else:
+                logger.debug('Case 4: quorum is bigger than needed')
+                yield from self.quorum_update(len(self.sync) - self.numsync, self.voters)
+        else:
+            safety_margin = self.quorum + self.numsync - len(self.voters | self.sync)
+            if self.numsync == self.sync_wanted and safety_margin > 0 and self.numsync > self.numsync_confirmed:
```

Comment:
Will we ever reach this case?

I mean, I would expect the `else` where this `if` is in to be triggered only if `numsync` != `sync_wanted`.

Is there anything that I'm missing?